src/psyclone/psyGen.py

```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+                 or readwrite argument.
+        :rtype: string
+        '''
+        return self._access
+
+    @access.setter
+    def access(self, new_access):
+        '''
+        :param str new_access: New value of the access attribute, can be:
+                               'local', 'external', 'read_arg', 'write_arg'
+                               or 'readwrite_arg'.
+        '''
+        if new_access not in ('local', 'external', 'read_arg', 'write_arg',
+                              'readwrite_arg'):
+            raise InternalError("Symbol access attribute can only be: "
+                                "'local', 'external', 'read_arg', "
+                                "'write_arg' or 'readwrite_arg'.")
+        self._access = new_access
 
     def __str__(self):
-        return (self.name + "<" + self.datatype + "," + str(self.dimensions) +
-                "," + self.kind + ">")
+        return (self.name + "<" + self.datatype + ", " + str(self.shape) +
+                ", " + self.access + ">")
 
 
 class SymbolTable(object):
     '''
-    Encapsulates the symbols table and provides methods to declare new symbols
-    and look up existing symbols. It is implemented as a monolithic scope
-    symbol table.
+    Encapsulates the symbol table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a single scope
+    symbol table (nested scopes not supported).
     '''
     def __init__(self):
+        # Dict of Symbol objects with the symbol names as keys.
         self._symbols = {}
+        # Ordered list of the arguments.
+        self._argument_list = []
 
-    def declare(self, name, datatype, dimensions, kind):
+    def declare(self, name, datatype, shape, access):
         '''
-        Declare a new symbol in the symbols table.
+        Declare a new symbol in the symbol table.
 
-        :param name: Name of the symbol.
-        :type name: string
-        :param datatype: Data type of the symbol.
-        :type datatype: string
-        :param dimensions: Dimensions of the symbol (0 represents an scalar
-                           symbol).
-        :type dimensions: list of integers
-        :param kind: Information about the variable declaration attributes.
-        :type kind: string
+        :param str name: Name of the symbol.
+        :param str datatype: Datatype of the symbol.
+        :param shape: Shape of the symbol (empty list represents an scalar
+                      symbol).
+        :type shape: list of integers
+        :param str access: Information about the variable declaration
+                           attributes.
         '''
         if name in self._symbols:
-            raise InternalError("Multiple definition of symbol: '{0}'."
-                                "".format(name))
-        else:
-            self._symbols[name] = Symbol(name, datatype, dimensions, kind)
+            raise InternalError("Symbol table already contains a symbol with"
```

Comment:
Thanks for improving the error message. On reflection I don't think this should be an InternalError but I'm not sure what it should be. Maybe ValueError or KeyError?

Reply:
I used KeyError